pkg/pulumiyaml/analyser.go

```diff
@@ -1098,13 +1104,32 @@ func (tc *typeCache) typeConfig(r *Runner, node configNode) bool {
 	return true
 }
 
-// checks for config type compatibility
-func unifyConfigType(a, b schema.Type) (schema.Type, bool) {
-	a, b = codegen.UnwrapType(a), codegen.UnwrapType(b)
-	if a.String() == b.String() {
-		return a, true
-	} else if (a == schema.IntType && b == schema.NumberType) || (b == schema.IntType && a == schema.NumberType) {
+// Checks for config type compatibility between an already registered/ old config key with a new key.
+// We allow Int to fall back to Number.
+// We also allow a new non-String key to fall back to String
+// since parsing CLI-set config types is a best effort and may actually be intended as a String.
+// If the new key is a String with a defined value, and the old key is not, then we
+// parse the new key to determine if it can fall back to the old key's type
+// (i.e. "16" would be valid as an Int but "foo" would not.)
+// TODO: remove the last case once `configuration` is deprecated.
```

Comment:
I think this reads clearer, but you are free to ignore me here.

```suggestion
// Config types are compatible if
// - They are the same type.
// - We are assigning an integer to a number.
// - We are assigning any type to a string.
// - We are assigning a string to some type T *and* the string can be unambiguously parsed into T.
// TODO: remove the last case once `configuration` is deprecated.
```

Reply:
LGTM. Although "assigning" might be a bit ambiguous, but I think the test cases help clarify it as well.